# SANS Publications

Rempel, A. and Wittler, R.: [SANS serif: alignment-free, whole-genome based phylogenetic reconstruction](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab444/6300510). Bioinformatics. (2021).

Wittler, R.: [Alignment- and reference-free phylogenomics with colored de Bruijn graphs](https://pub.uni-bielefeld.de/download/2942421/2942423/s13015-020-00164-3.wittler.pdf).
Algorithms for Molecular Biology. 15: 4 (2020).

Wittler, R.: [Alignment- and reference-free phylogenomics with colored de Bruijn graphs](http://drops.dagstuhl.de/opus/volltexte/2019/11032/pdf/LIPIcs-WABI-2019-2.pdf).
In: Huber, K. and Gusfield, D. (eds.) Proceedings of WABI 2019. LIPIcs. 143, Schloss Dagstuhl--Leibniz-Zentrum fuer Informatik, Dagstuhl, Germany (2019).


# Splitstree

Huson, D. H. and Bryant, D.: [Application of Phylogenetic Networks in Evolutionary Studies](https://doi.org/10.1093/molbev/msj030), Mol. Biol. Evol., 23(2):254-267 (2006).

# Example dataset

Shapiro, B. J., Levade, I., Kovacikova, G., Taylor, R. K. and Almagro-Moreno S.: [Origins of pandemic Vibrio cholerae from environmental gene pools](https://doi.org/10.1038/nmicrobiol.2016.240), Nat. Microbiol., 2(3):16240 (2017).

preprocessed with [ORFfinder](https://ftp.ncbi.nlm.nih.gov/genomes/TOOLS/ORFfinder/)

